Fixing package module related issues.

nhosoi opened the pull request:
- Gathering package facts with package_facts first.
- Using ansible_facts.packages['rsyslog'][0].version for the rsyslog version.
- Renaming a variable rsyslog_packages with rsyslog_extra_packages to make it clearer. With this variable, customers are allowed to install extra rsyslog packages required for their custom configurations if any.
- Renaming a variable backupdir with rsyslog_backup_dir. Pre-existing config files are copied to the directory.  Default to /tmp/rsyslog.d-DATETIME.
- Renaming a pair of variables rsyslog_role_packages and rsyslog_role_rules with __rsyslog_packages and __rsyslog_rules, respectively.  Note: the vars start with `__` since they are internal.  By the same token, the underneath sub rule names and their items are also changed to start with __, e.g.,
      rsyslog_common_rules --> __rsyslog_common_rules
      rsyslog_conf_global_options --> __rsyslog_conf_global_options
- The internal variables are moved from defaults/main.yaml to vars/main.yaml.
- Fixing the relative path from role_path in roles/rsyslog/roles/*_roles/*/tasks/main.yaml.
- Removing set_test from CI tests.
- To call include_role, instead of setting variables to pass with set_fact, set them with "vars" to make them local in the scope.
- If rsyslog is not found, e.g., in case of the check mode, set 0.0.0 to __rsyslog_version.

Fixing travis-ci
- forcing to install molecule < 3.
- fixing yamllint and ansible-lint errors excepot 'state: latest' for the package module, lines longer than 160 chars, and commands should not change things if nothing needs doing.
- replacing var use_rsyslog_image with rsylog_in_image and set it to false by default.

nhosoi commented:
Note to myself...
This PR is depend on https://github.com/linux-system-roles/logging/pull/84/.
Once pr/84 is merged, solve the conflicts and merge this PR.

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -1,19 +1,19 @@
-    name: "{{ lookup('flattened', rsyslog_base_packages, rsyslog_tls_packages if rsyslog_pki else [], rsyslog_packages) }}"
+    name: "{{ item }}"
     state: latest
+  loop: "{{ _rsyslog_packages | flatten | d('') }}"
```
Why this change? https://github.com/oasis-roles/meta_standards#ansible-best-practices says: Avoid calling the package module iteratively with the {{ item }} argument, as this is impressively more slow than calling it with the line name: "{{ foo_packages }}". The same can go for many other modules that can be given an entire list of items all at once.

nhosoi replied:
Understood...   I introduced the loop to check whether each package is already installed or not here [1].  (Otherwise, I got an error from package/yum, so I had to put [2] to ignore it.  I thought it'd be better getting rid of [2].)  What'd be the best way to handle that kind of contradictory issue?  Do you recommend just passing the list to the `package` module and setting the ignore error?
[1] https://github.com/linux-system-roles/logging/pull/87/files#diff-ed5a08a2e61430dc18e2d3820cbd6f32R13
[2] https://github.com/linux-system-roles/logging/pull/87/files#diff-ed5a08a2e61430dc18e2d3820cbd6f32L9

pcahyna replied:
What error did you get?

pcahyna replied:
(and yes, using `ignore_errors` like [2] is usually not a very good idea. But other roles and the best practices have been using package list without a loop and without `ignore_errors`. If there is a problem with applying the best practices, please investigate it and report it, instead of doing things differently.)

nhosoi replied:
Hmmm, even though I reverted this change, I cannot reproduce the error...  Maybe, the cause was something else.  Sorry for the noise.  I keep reverting this change for now.

nhosoi replied:
Note: I reverted the code pointed out by @pcahyna.

pcahyna commented on roles/rsyslog/defaults/main.yaml:
```diff
@@ -67,10 +67,10 @@ rsyslog_base_packages: ['rsyslog']
 #
 # List of additional rpm packages to install.
-rsyslog_packages: []
+rsyslog_additional_packages: []
```
Do you use this anywhere? IMO this only complicates the role and breaks abstraction. If installing extra packages is needed for some functionality, there should be a parameter for the functionality and the role should know what packages to install, instead of requiring the user to know this (it may be different for different supported distributions and versions).
If we decide to keep this functionality, I suggest to follow the example of OASIS, where it is suggested to be named extra_packages instead of additional_packages: https://github.com/oasis-roles/meta_standards#vars-vs-defaults
@robled @greg-hellings @seandst have you found such a variable useful in practice?

nhosoi replied:
I understand the issue, but I reluctant to remove this variable...
The rsyslog configuration is quite flexible and it supports these plugins.  Some of them are not installed by the current logging role.  But customers may want to manually configure it in their environment.  To cover such a use case, this rsyslog_additional_packages might be useful.
```
rsyslog-crypto-8.1911.0-3.el8.x86_64.rpm
rsyslog-elasticsearch-8.1911.0-3.el8.x86_64.rpm
rsyslog-gnutls-8.1911.0-3.el8.x86_64.rpm
rsyslog-gssapi-8.1911.0-3.el8.x86_64.rpm
rsyslog-kafka-8.1911.0-3.el8.x86_64.rpm
rsyslog-mmaudit-8.1911.0-3.el8.x86_64.rpm
rsyslog-mmjsonparse-8.1911.0-3.el8.x86_64.rpm
rsyslog-mmkubernetes-8.1911.0-3.el8.x86_64.rpm
rsyslog-mmnormalize-8.1911.0-3.el8.x86_64.rpm
rsyslog-mmsnmptrapd-8.1911.0-3.el8.x86_64.rpm
rsyslog-mysql-8.1911.0-3.el8.x86_64.rpm
rsyslog-pgsql-8.1911.0-3.el8.x86_64.rpm
rsyslog-relp-8.1911.0-3.el8.x86_64.rpm
rsyslog-snmp-8.1911.0-3.el8.x86_64.rpm
```

pcahyna replied:
I see. Still, can't you introduce more generic variables to control whether a specific functionality will be enabled and the plugin installed? I have tho concerns:
- the name of a plugin implementing a specific functionality can change, the role should isolate the user from such changes, but if the user has to know the exact names of packages, it does not fulfil this function,
- presumably to make use of a plugin, sometimes it could be necessary to configure it in addition to installing it (is that the case?), and this again calls for a more high-level approach, because when the role just installs given plugins, it will not help the user with configuring them.
If you decide to keep the variable, I would suggest to at least rename it to the OASIS convention.

nhosoi replied:
I renamed the misleading var name rsyslog_additional_packages with rsyslog_extra_packages as suggested.

pcahyna commented on roles/rsyslog/roles/input_roles/basics/defaults/main.yaml:
```diff
@@ -12,6 +12,7 @@ rsyslog_unprivileged: false
 #       Note: unles rsyslog_send_over_tls_only is set to true, insecure network connection
 #       is also allowed.
 rsyslog_capabilities: []
+rsyslog_basics_packages: ['rsyslog-gnutls']
```
is the user supposed to override this? If not, it should go to vars, not defaults, and be prefixed with double underscore.

nhosoi replied:
Rsyslog supports openssl, as well.  But it looks the plugin is not included in the RHEL/CentOS distribution.  So, there's no chance to override for now.  I will follow your advice.  Thanks!

nhosoi replied:
Note: removed the var.

pcahyna commented on roles/rsyslog/roles/input_roles/basics/tasks/main.yaml:
```diff
@@ -1,7 +1,7 @@
 - name: Set basics facts
   set_fact:
-    rsyslog_role_packages: rsyslog-gnutls
+    rsyslog_role_packages: "{{ rsyslog_basics_packages }}"
```
shouldn't it be called just `rsyslog_packages`? Why the `role` in the name? Also, why not drop the `set_fact` and rename `rsyslog_basics_packages` to `rsyslog_packages` (or `__rsyslog_packages` if the variable is supposed to be internal, and I suspect it is)?

nhosoi replied:
The var rsyslog_role_packages is used here [1].  They need to match.  The current rsyslog_role_packages could be set in each input and output sub-roles.  So, I think we should replace rsyslog_role_packages with __rsyslog_packages to make it internal, and replace each rsyslog_something_packages with __rsyslog_something_packages?
[1] https://github.com/linux-system-roles/logging/blob/master/roles/rsyslog/tasks/deploy.yaml#L2-L8

nhosoi replied:
I modified the code like this
https://github.com/linux-system-roles/logging/pull/87/files#diff-532de15639f14a0202b8cd2880a8bd4cR4-R10
with the comment:
- To call include_role, instead of setting variables to pass with set_fact, put them in the block with vars to make them local in the scope.

Please note that I made the same change to all the include_role.

pcahyna replied:
You don't need a block. Put vars: directly on the include_role statement.

nhosoi replied:
Thanks, @pcahyna.  Fixed.

pcahyna commented on roles/rsyslog/roles/input_roles/ovirt/tasks/main.yaml:
```diff
@@ -1,12 +1,12 @@
 - name: Set oVirt facts
   set_fact:
-    rsyslog_role_packages: "{{ rsyslog_ovirt_prereq_packages|flatten|d([]) }}+{{ rsyslog_ovirt_packages|flatten|d([]) }}"
+    rsyslog_role_packages: "{{ rsyslog_ovirt_prereq_packages | d([]) + rsyslog_ovirt_packages | d([]) }}"
```
See above - can it be renamed to `__rsyslog_packages` and moved to vars, eliminating `rsyslog_ovirt_packages` ?
What will happen if one uses both the `basics` and `ovirt` subroles? Won't the second `set_fact` clobber the result of the former? is that expected?

pcahyna replied:
maybe I misunderstood - I see there are `rsyslog_base_packages` and `rsyslog_role_packages` variables with different functions - why is it the case?

nhosoi replied:
The rsyslog_role_packages are taken care in roles/rsyslog/tasks/deploy.yaml vir each deploy.yaml following tasks_from line (e.g., https://github.com/linux-system-roles/logging/pull/87/files#diff-32e44b317e7ab0e0016656a25254fa0bL10, here).  So, ovirt packages are installed immediately after this set_fact and the same happens for basics, too.

> Won't the second set_fact clobber the result of the former?

So, the answer is "no".  It does not clobber the result of the former, but the value of rsyslog_role_packages is used immediately before being clobbered.

nhosoi replied:
https://github.com/linux-system-roles/logging/pull/87#discussion_r381552024 is applied here, as well.

nhosoi replied:
Note: var name `rsyslog_role_packages` was renamed to `__rsyslog_packages` and the var is set with instead of `set_fact` but in `var:`.

pcahyna commented on roles/rsyslog/roles/input_roles/viaq-k8s/tasks/main.yaml:
```diff
@@ -1,10 +1,10 @@
 - name: Set Viaq-k8s facts
   set_fact:
-    rsyslog_role_packages: "{{ rsyslog_viaq_k8s_packages|flatten }}"
+    rsyslog_role_packages: "{{ rsyslog_viaq_k8s_packages }}"
```
see above...

nhosoi replied:
#87 (comment) is applied here, as well.

nhosoi replied:
Note: var name `rsyslog_role_packages` was renamed to `__rsyslog_packages` and the var is set with instead of `set_fact` but in `var:`.

pcahyna commented on roles/rsyslog/roles/input_roles/viaq/tasks/main.yaml:
```diff
@@ -1,10 +1,10 @@
 - name: Set Viaq facts
   set_fact:
-    rsyslog_role_packages: "{{ rsyslog_viaq_prereq_packages|flatten|d([]) }}+{{ rsyslog_viaq_packages|flatten|d([]) }}"
+    rsyslog_role_packages: "{{ rsyslog_viaq_prereq_packages + rsyslog_viaq_packages }}"
```
see above

nhosoi replied:
#87 (comment) is applied here, as well.

nhosoi replied:
Note: var name `rsyslog_role_packages` was renamed to `__rsyslog_packages` and the var is set with instead of `set_fact` but in `var:`.

pcahyna commented on roles/rsyslog/roles/output_roles/elasticsearch/tasks/main.yaml:
```diff
@@ -53,7 +53,7 @@
 - name: Set Elasticsearch facts
   set_fact:
-    rsyslog_role_packages: "{{ rsyslog_elasticsearch_packages | flatten([]) }}"
+    rsyslog_role_packages: "{{ rsyslog_elasticsearch_packages }}"
```
see above

nhosoi replied:
#87 (comment) is applied here, as well.

nhosoi replied:
Note: var name `rsyslog_role_packages` was renamed to `__rsyslog_packages` and the var is set with instead of `set_fact` but in `var:`.

pcahyna commented on tests/tests_listen.yml:
```diff
@@ -30,6 +30,8 @@
         name: lsof
         state: present
+      when:
+        - not 'lsof' in ansible_facts.packages
```
I know this makes the test a bit faster, but is it worth complicating the code and making it more fragile because of this optimization?

nhosoi replied:
If it's not preferable, I'm removing the line 33 and 34.  I looked into other linux-system-roles and thought it's recommended. :)

nhosoi replied:
Note: `when: not 'lsof' in ansible_facts.packages` was removed;

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -1,19 +1,19 @@
-  package:
-    list: rsyslog
-  register: rsyslog_version
+  set_fact: rsyslog_version=ansible_facts.packages['rsyslog'][0].version | default('') | int
```
Doesn't the block below need adjusting? Also, does it need to be a fact instead of a variable local to the block?

nhosoi replied:
Fixed.

nhosoi commented:
Hi, @pcahyna, @sradco, @richm, could you please review this pr one more time?  I've updated the patch to eliminate set_fact for include_role.

Actually, I have another patch to make the logging role more compliant to the best practices [1].  (I think this one is rather straightforward although it is huge... :)  It's still my private pr, but once this pr/87 is merged, I'm going to submit an official pull request from it.  Thanks!
[1] https://github.com/nhosoi/logging/pull/1

pcahyna commented on roles/rsyslog/README.md:
```diff
@@ -338,7 +338,7 @@ WARNING: Pre-existing rsyslog.conf and configuration files in /etc/rsyslog.d are
 Describing how the configuration files are defined to be deployed using the viaq case.
 
-Viaq configuration files are defined in {{ rsyslog_viaq_rules }} in /roles/input_rols/viaq/defaults/main.yaml.  The set is made from the generic modules{rsyslog_conf_global_options, rsyslog_conf_local_modules, rsyslog_conf_network_modules, rsyslog_conf_common} and viaq specific configurations.
+Viaq configuration files are defined in {{ __rsyslog_viaq_rules }} in /roles/input_rols/viaq/defaults/main.yaml.  The set is made from the generic modules{rsyslog_conf_global_options, rsyslog_conf_local_modules, rsyslog_conf_network_modules, rsyslog_conf_common} and viaq specific configurations.
```
If you are prefixing the variable name with underscores, it is internal and should it be documented then at all? (maybe it should, but in some developer/internal focused documentation, not in user documentation.) also, there is a typo in `input_rols`

nhosoi replied:
:p  thanks!

nhosoi replied:
The typo input_rols was fixed.
Documentation for the internal vars will be added in a separate pr.

pcahyna commented on tests/tests_files_forwards.yml:
```diff
@@ -51,27 +53,23 @@
     - name: Check if the files config exists
       assert:
-        that: stat_output_files_system.stat.exists == True
+        that: stat_output_files_system.stat.exists == true
```
why compare a boolean to true? won't it work to just use `stat_output_files_system.stat.exists` ?

pcahyna replied:
this applies to other places as well - but please test it first

pcahyna commented on roles/rsyslog/defaults/main.yaml:
```diff
@@ -291,60 +281,29 @@ rsyslog_weight_map:
-rsyslog_common_rules:
+# configuration in a special format. This lis specifies default ``rsyslogd``
+# configuration enabled in the role.
+__rsyslog_common_rules:
```
If you are prepending underscores to variable names, does it mean that the variables are internal and therefore don't need to be settable by users? - but then they belong to vars and not defaults, no?

Also I would say that this diverges from the topic of fixing package module related issues.

nhosoi replied:
> If you are prepending underscores to variable names, does it mean that the variables are internal and therefore don't need to be settable by users? - but then they belong to vars and not defaults, no?

That's correct.  They should belong to vars.

> Also I would say that this diverges from the topic of fixing package module related issues.

+1.  May I stop this here and move to another PR for the internal var fixes (including docs)?

pcahyna replied:
yes please, finish the package related changes in this PR and continue variable renaming (if not related to the package module changes) in another PR.

pcahyna replied:
and please rename `rsyslog_packages` to `rsyslog_extra_packages` instead of to `rsyslog_additional_packages`, the former seems more idiomatic to me.

nhosoi replied:
Done.

nhosoi commented:
@pcahyna, could you please help me?  With the changes I made today, the molecule (ansible lint) started failing like this [1].  But I could not find out where the failure "{'lint': ['must be of string type']}" comes from... :(

https://travis-ci.org/linux-system-roles/logging/jobs/653146080?utm_medium=notification&utm_source=github_status

pcahyna commented:
> @pcahyna, could you please help me? With the changes I made today, the molecule (ansible lint) started failing like this [1]. But I could not find out where the failure "{'lint': ['must be of string type']}" comes from... :(
> 
> https://travis-ci.org/linux-system-roles/logging/jobs/653146080?utm_medium=notification&utm_source=github_status

I am about to leave. Please ask @richm

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -1,43 +1,31 @@
-    list: rsyslog
-  register: rsyslog_version
+  set_fact:
+    __rsyslog_version: "{{ ansible_facts.packages['rsyslog'][0].version }}"
```
Hmm I thought about a problem, what if rsyslog is not installed here? With @vcrhonek we have been discussing a similar problem in the timesync role.

nhosoi replied:
> Hmm I thought about a problem, what if rsyslog is not installed here? 

Could it happen even if rsyslog is installed with the package module just above this task?

pcahyna replied:
yes, think about check mode.

nhosoi replied:
> yes, think about check mode.

I see.  Do you recommend to add `check_mode: no` to the install packages tasks?

pcahyna replied:
No, this would defeat the purpose of using check mode, which is to not make changes to the system.

nhosoi replied:
Indeed...  We can do either 1) quit with no such package error here or 2) pretend having an rsyslog package and keep running since it is a dry-run.  Which would be better?

pcahyna replied:
I would say some version of 2) since I don't think we should fail in check mode if we would not fail in normal mode. We could even skip the rest of the role since we properly reported changed status.

pcahyna replied:
Here's a fix for a similar problem in timesync: https://github.com/linux-system-roles/timesync/pull/60

nhosoi replied:
Thank you so much, @pcahyna.  I tested this diff and verified it works as expected.
https://github.com/nhosoi/logging/compare/best_practices...nhosoi:package_version

Once pr/89 is done, I'm opening a pr with it.
https://github.com/linux-system-roles/logging/pull/89

pcahyna replied:
@nhosoi I did not intend to say that you must follow the patch taken in timesync. Rather we should determine which approach is better and use it consistently.
(yet another approach would be to not support check mode on first run of the role until Ansible 2.7 is deprecated and we can solve the problem much more elegantly using the default filter).

richm commented:
Looks like we are now using molecule v3?
```
>pwd
/home/rmeggins/linux-system-roles/logging
>git pull
up-to-date
>python -mvenv .venv
>. .venv/bin/activate
>pip install molecule docker
>molecule --version
molecule 3.0.1
   ansible==2.9.5 python==3.7
```
and they changed the molecule config?  https://github.com/ansible-community/molecule/blob/master/docs/configuration.rst
```
Lint

Starting with v3, Molecule handles project linting by invoking and external lint commands as exemplified below.

The decision to remove the complex linting support was not easily taken as we do find it very useful. The issue was that molecule runs on scenarios and linting is usually performed at repository level.

It makes little sense to perform linting in more than one place per project. Molecule was able to use up to three linters and while it was aimed to flexible about them, it ended up creating more confusions to the users. We decided to maximize flexibility by just calling an external shell command.

lint: |
    yamllint .
    ansible-lint
    flake8

The older format is no longer supported and you have to update the molecule.yml when you upgrade. If you don't want to do any linting, it will be enough to remove all lint related sections from the file.

# old v2 format, no longer supported
lint:
  name: yamllint
  enabled: true
...
```

richm commented:
@nhosoi so what I would suggest is to update the molecule config for v3 - you can test it locally as I have above
If that seems too difficult, we can change the tests - travis - whatever to use molecule v2.  For example:
```
>pip install 'molecule<2'
...
>molecule --version
molecule, version 2.22
```

richm commented:
the molecule config file is `molecule/default/molecule.yml`

nhosoi commented:
> molecule/default/molecule.yml

Thanks a lot, @richm.  By upgrading the molecule, I see more errors from ansible-lint... :)  @richm, @pcahyna, is it ok to fix them in this pr/87?  Or shall I just merge this pr with the failed travis and submit a new pull request for the lint fixes?

richm commented:
I think the molecule config/lint fixes should be in a separate PR - that's a lot of stuff . . .

nhosoi commented:
Thanks, @richm!
> I think the molecule config/lint fixes should be in a separate PR - that's a lot of stuff . . .

@pcahyna, is it ok to merge this pr without the second commit "Fixing travis-ci"?
https://github.com/linux-system-roles/logging/pull/87/commits/c772da4a0045e0f45608a28406704db7a2c6d87d

pcahyna commented:
@i386x @tyll @vcrhonek @xjezda00 this is something to be taken account of in our template: https://github.com/linux-system-roles/logging/pull/87#issuecomment-589298865
FYI @robled @greg-hellings @seandst

pcahyna commented on roles/rsyslog/tasks/deploy.yaml:
```diff
@@ -1,20 +1,23 @@
   package:
-    name: "{{ rsyslog_role_packages }}"
+    name: "{{ __rsyslog_packages }}"
     state: latest
```
I propose to use `present`, not `latest`. Most of our roles use this, so let's be consistent. @robled @greg-hellings @seandst do you have an opinion here?
If we want to be able to use `latest`, we can introduce a _xxx_`_packages_state` variable, like here:

https://github.com/geerlingguy/ansible-role-apache/blob/77f39119bc12fd839d16a55a19354d5c159041fa/defaults/main.yml#L58

nhosoi replied:
Leaving the state as latest to make sure the newer packages are installed/updated if available.
Leaving the conversation opened.

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -1,44 +1,25 @@
   package:
-    name: "{{ lookup('flattened', rsyslog_base_packages, rsyslog_tls_packages if rsyslog_pki else [], rsyslog_packages) }}"
+    name: "{{ (__rsyslog_base_packages) + (__rsyslog_tls_packages if rsyslog_pki else []) + (rsyslog_additional_packages) }}"
     state: latest
```
see the note above regarding `latest`

nhosoi replied:
I'm wondering when the system has the older rsyslog package is installed and there is a newer rsyslog in the repository, what happens if the state is present vs. latest.  Can we make sure we could install the latest package?

pcahyna replied:
It hasn't been a problem for other system roles apparently.

nhosoi replied:
Ok, it's guaranteed to be upgraded with `present`.  Reading this doc, I was not sure it was. :)
https://docs.ansible.com/ansible/latest/modules/package_module.html#parameter-state

I'm going to revert "latest" changes...

richm replied:
> what happens if the state is present vs. latest

If the rsyslog package does not exist, then both `state: present` and `state: latest` will install the latest package available.

If the rsyslog package does exist, then `state: present` is a no-op, and `state: latest` will do an upgrade to the latest package available, if there is a newer one available.

richm replied:
Also - if we do not hardcode `state: latest` in the role, then we need some other mechanism like https://github.com/geerlingguy/ansible-role-apache/blob/77f39119bc12fd839d16a55a19354d5c159041fa/defaults/main.yml#L58 - also note that if we do this, we should be consistent among *all* of the system roles - that is, every system role will need to have a `ROLE_packages_state: present` in the `defaults/main.yml` - we'll also need to add this to the README.md, the template, and the best practices guide.

nhosoi replied:
> > what happens if the state is present vs. latest
> If the rsyslog package does exist, then `state: present` is a no-op, and `state: latest` will do an upgrade to the latest package available, if there is a newer one available.

Thanks, @richm.  Hmmm, then I'd like to keep `latest` for rsyslog package install/upgrade.  The logic is also used for updating security related packages.  So, I think we need to make sure it updates to the latest.

pcahyna replied:
Ok, keep it for now, but we should set a common standard for all the roles, because I don't see why logging should be different.

nhosoi replied:
Thanks, @pcahyna.  Just out of curiosity, there's no requirement for other roles to _upgrade_ existing packages?

pcahyna replied:
I am not aware of such a requirement. As you can see above, the default for many popular roles on Galaxy is to not update.

pcahyna commented on tests/tests_listen.yml:
```diff
@@ -29,25 +28,36 @@
       package:
         name: lsof
-        state: present
+        state: latest
```
do you need it? I suggest to omit this change.

nhosoi replied:
I don't care the version of lsof...  So, reverting this. :)

pcahyna commented:
> > molecule/default/molecule.yml
> 
> Thanks a lot, @richm. By upgrading the molecule, I see more errors from ansible-lint... :) @richm, @pcahyna, is it ok to fix them in this pr/87? Or shall I just merge this pr with the failed travis and submit a new pull request for the lint fixes?

@i386x @tyll have you seen such problems in other roles? If not, why not? We are using molecule elsewhere.

richm commented:
> > > molecule/default/molecule.yml
> > 
> > 
> > Thanks a lot, @richm. By upgrading the molecule, I see more errors from ansible-lint... :) @richm, @pcahyna, is it ok to fix them in this pr/87? Or shall I just merge this pr with the failed travis and submit a new pull request for the lint fixes?
> 
> @i386x @tyll have you seen such problems in other roles? If not, why not? We are using molecule elsewhere.

You would only see this if you are actively working on a role, or actively working on a PR, within the past few days, after the version of molecule on pypi was updated to v3.

tyll commented:
> > > molecule/default/molecule.yml
> > 
> > 
> > Thanks a lot, @richm. By upgrading the molecule, I see more errors from ansible-lint... :) @richm, @pcahyna, is it ok to fix them in this pr/87? Or shall I just merge this pr with the failed travis and submit a new pull request for the lint fixes?
> 
> @i386x @tyll have you seen such problems in other roles? If not, why not? We are using molecule elsewhere.

What are the acutal errors, I saw this:
https://travis-ci.com/linux-system-roles/network/jobs/289283294#L460 in https://github.com/linux-system-roles/network/pull/153

```
verifier:
  lint:
    name: flake8
  name: testinfra

ERROR: Failed to pre-validate.

{'lint': ['must be of string type']}
ERROR: InvocationError for command /bin/bash .travis/runsyspycmd.sh .tox/env-3.5/bin/python /usr/bin/python3 molecule lint -s default (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   molecule: commands failed
```

Sounds like it could be the issue.

richm commented:
> {'lint': ['must be of string type']}

Yes, this is the issue - this means the molecule.yml config needs to be upgraded to support molecule v3.

tyll commented:
> > {'lint': ['must be of string type']}
> 
> Yes, this is the issue - this means the molecule.yml config needs to be upgraded to support molecule v3.

What should it look like now?

```
 lint: |
    yamllint --config-file molecule/default/yamllint.yml .
```

checks also everything in the tox directory...

richm commented:
> > > {'lint': ['must be of string type']}
> > 
> > 
> > Yes, this is the issue - this means the molecule.yml config needs to be upgraded to support molecule v3.
> 
> What should it look like now?
> 
> ```
>  lint: |
>     yamllint --config-file molecule/default/yamllint.yml .
> ```
> 
> checks also everything in the tox directory...

I don't know - here is the document about the new config:  https://github.com/ansible-community/molecule/blob/master/docs/configuration.rst

@nhosoi - did you run into any problems with the checker descending into venv or tox directories, and if so, how did you disable that?

nhosoi commented:
> > > > {'lint': ['must be of string type']}
> > > 
> > > 
> > > Yes, this is the issue - this means the molecule.yml config needs to be upgraded to support molecule v3.
> > 
> > 
> > What should it look like now?
> > ```
> >  lint: |
> >     yamllint --config-file molecule/default/yamllint.yml .
> > ```
> > 
> > 
> > checks also everything in the tox directory...
> 
> I don't know - here is the document about the new config: https://github.com/ansible-community/molecule/blob/master/docs/configuration.rst
> 
> @nhosoi - did you run into any problems with the checker descending into venv or tox directories, and if so, how did you disable that?

Ahhh, I was not running yamllint correctly.  Fixed it.  I listed subdirectories to test instead of '.'.

pcahyna commented:
@nhosoi I think you don't need to worry about Travis tests, because we have the same problems in other roles and we need to solve them together.

nhosoi commented:
> @nhosoi I think you don't need to worry about Travis tests, because we have the same problems in other roles and we need to solve them together.

@pcahyna, ok.  Then, what shall I do?  Is it ok to merge this pr or further reviews/fixes are needed?  If it's ok to merge, is it with Fixing travis-ci https://github.com/linux-system-roles/logging/pull/87/commits/9d606287ec06e5a02e6a98c69d7664f768047fcc, or without it?  Please note that the patch contains the fixes reported by the travis tests, as well.  So, if it's allowed, I'd like to merge both of the patches https://github.com/linux-system-roles/logging/pull/87/commits/a27c2c0f8eb50edababb400cd8416f8cab1a97c7 and https://github.com/linux-system-roles/logging/pull/87/commits/9d606287ec06e5a02e6a98c69d7664f768047fcc together.  Thanks.

pcahyna commented:
Sure, if you have the issue already fixed, keep the fix, I meant that you don't need to work further on a fix sepcifically for logging now.

nhosoi commented:
> Sure, if you have the issue already fixed, keep the fix, I meant that you don't need to work further on a fix sepcifically for logging now.

@pcahyna, thank you!  Now, may I merge this patch?  Or further reviews are needed?  If it's ok to merge, could some of you please approve this pr?  Thanks! > @pcahyna, @richm, @sradco

pcahyna commented on roles/rsyslog/defaults/main.yaml:
```diff
@@ -291,60 +286,29 @@ rsyslog_weight_map:
-# details. This list should be used for configuration by other Ansible roles.
-rsyslog_dependent_rules: []
-
-# rsyslog_.*_rules
```
how is the deleted block above related to the subject of this PR?

nhosoi replied:
Since I found these "rules" are not used anywhere, I cleaned this up.  Do you want me to move this to a new pr?

pcahyna commented on roles/rsyslog/defaults/main.yaml:
```diff
@@ -383,11 +347,57 @@ rsyslog_conf_local_modules:
+                  if ("tls" in rsyslog_capabilities)
+                  else "absent" }}'
+
+# __rsyslog_conf_common_defaults
```
this is a big change, what is the relation to the rest of the PR?

nhosoi replied:
It's not...  I'm removing this section.  Sorry for the noise. :(

pcahyna commented:
@nhosoi I looked at the PR again and found two big changes that seem to be unrelated to the PR description, are they intended?

nhosoi commented:
@pcahyna, they were part of cleaning up, but they should not be in this pr.  Sorry, I'm removing them.

pcahyna commented:
@nhosoi can you please review the PR again for unintended changes yourself?

richm commented:
please revert the change to ` molecule/default/molecule.yml` - instead, change your `tox.ini` like this:
```
--- a/tox.ini
+++ b/tox.ini
@@ -176,7 +176,7 @@ commands =
 deps =
     {env:LSR_ANSIBLE_DEP:ansible}
     docker
-    molecule
+    molecule<3
     selinux
     -rmolecule_extra_requirements.txt
```
The plan is to use molecule v2 until we complete all of the work necessary to port all of the system roles to v3

pcahyna commented:
@richm do we want to save that change in a branch, since it seems to work and could be used as a basis for further changes?

richm commented:
> @richm do we want to save that change in a branch, since it seems to work and could be used as a basis for further changes?

Yes, good idea

nhosoi commented:
> please revert the change to ` molecule/default/molecule.yml` - instead, change your `tox.ini` like this:
> 
> ```
> --- a/tox.ini
> +++ b/tox.ini
> @@ -176,7 +176,7 @@ commands =
>  deps =
>      {env:LSR_ANSIBLE_DEP:ansible}
>      docker
> -    molecule
> +    molecule<3
>      selinux
>      -rmolecule_extra_requirements.txt
> ```
> 
> The plan is to use molecule v2 until we complete all of the work necessary to port all of the system roles to v3

It seems there's no tox.ini in the logging roles... :(

richm commented:
> It seems there's no tox.ini in the logging roles... :(

yet another enhancement/fixup to bring this role into sync with the other roles . . . 

for now - looks like molecule is installed in a couple of places:
```
.travis.yml:  - pip install molecule docker
README.md:$ pip install molecule docker
```
change both of these to `'molecule<3'`
also - if you want to test locally, and you already have molecule v3, do this:
```
pip uninstall molecule
pip install 'molecule<3'
```

nhosoi commented:
@pcahyna,
> @nhosoi can you please review the PR again for unintended changes yourself?

I went through the pr and removed the 2 changes out of the scope of this pr, which you pointed out.  I also updated the commit comments to describe the patches.  Thanks.

nhosoi commented:
> > It seems there's no tox.ini in the logging roles... :(
> 
> yet another enhancement/fixup to bring this role into sync with the other roles . . .

@richm, I looked into the other roles.  (I don't plan to add it to this pr.  Please don't worry... :)  The network role has tox.ini, but not timesync and selinux.  I see tox.ini in template.  That means it's new and recommended to add?

richm commented:
> That means it's new and recommended to add?

It means we will fix it later, if necessary

sradco commented on roles/rsyslog/defaults/main.yaml:
```diff
@@ -322,29 +317,28 @@ rsyslog_host_rules: []
-# details. This lis specifies default ``rsyslogd`` configuration enabled in the
-# role.
-rsyslog_common_rules:
+# configuration in a special format. This lis specifies default ``rsyslogd``
```
@nhosoi list* instead of lis*

nhosoi replied:
Thanks.  Fixed.

sradco commented on roles/rsyslog/roles/README.md:
```diff
@@ -21,7 +21,7 @@ The `defaults` directory includes:
 The `tasks` directory includes 2 tasks file:
   - `main.yaml` - tasks for deploying the config files
-    This file is sets `rsyslog_role_packages` and `rsyslog_role_rules` and includes the task that deploys the files.
+    This file is sets `__rsyslog_packages` and `__rsyslog_rules` and includes the task that deploys the files.
```
@nhosoi This file sets. I think we can drop the "is".

nhosoi replied:
Thanks.  Fixed.

sradco commented on roles/rsyslog/roles/input_roles/basics/tasks/main.yaml:
```diff
@@ -1,10 +1,9 @@
+# Deploy configuration files
 - name: Install/Update basics role packages and generate example configuration files to role subdir
+  vars:
+    __rsyslog_packages: "{{ __rsyslog_basics_packages }}"
```
@nhosoi you don't need faltten?

nhosoi replied:
Basically, as long as one level list ['package0', 'package1', ...] is passed, flatten is not needed.  You might have noticed that the variables like `__rsyslog_basic_packages` start with `__` that means they are internal set in each vars/main.yaml files.  We do not allow users to set them.  So, it's guaranteed that the passed lists are one level.

But there is one exception. :)  We allow users to install/update extra rsyslog packages with rsyslog_extra_packages here.
https://github.com/linux-system-roles/logging/pull/87/files#diff-ed5a08a2e61430dc18e2d3820cbd6f32R3
As you suggested, we'd better flatten rsyslog_extra_packages.  For instance, if this kind of nested list is given, it fails in the task.  I'm adding the flatten filter to rsyslog_extra_packages.  Thanks!!
rsyslog_extra_packages: [['rsyslog-mysql'], ['rsyslog-kafka']]

pcahyna replied:
@nhosoi why should such a nested list appear there?

nhosoi replied:
> @nhosoi why should such a nested list appear there?

@pcahyna, if we allow users to set a variable, any input could come in, so we'd better be prepared, I thought.

pcahyna replied:
@nhosoi I don't think we should expect any input, we should expect correct input. (Will we flatten all the lists passed to the role? If not, why is the package list special? If yes, will we do it in all the roles? We should as the roles should be consistent.)

nhosoi replied:
@pcahyna, in my past experiences, I learned users could make any mistakes and ask supports for help; to reduce the calls, we are supposed to add any efforts as much as possible. :)  Let's assume a user put an invalid list.  I guess we need to choose one of these?
1. with flatten, logging role successfully works.
2. with some list inspection code (which does not exist now, I think), logging role fails with a descriptive error.
3. without any support, logging role fails with a cryptic error...
@richm, @sradco, can I have your thoughts?

richm replied:
Having some sort of input normalization/validation for this parameter establishes a precedent, that we will have input normalization/validation for _all_ input parameters for _all_ system roles.  I don't think we want that.  I don't think we need that, except in very specific cases where:
* it is easy for users to make a mistake
* the mistake could be very costly (e.g. some sort of destructive operation that loses data)

I don't think this qualifies, so I would prefer not to have any extra code that makes this more complex.

pcahyna replied:
I don't think we need a decision now. Please do not add new instances of `flatten`, but keep the existing ones. This is a topic for standards discussion.

pcahyna replied:
@richm I agree. Moreover, input normalization for cases where the mistake could be very costly is even less justified, because it could convert user input to something that the user has not intended (OTOH input validation would be justified.)

nhosoi replied:
@pcahyna, @richm, thanks for your discussions.  Yes, it'd be nice to have some standards for handling possible user mistakes.

sradco commented on roles/rsyslog/roles/input_roles/files/tasks/cleanup.yaml:
```diff
@@ -1,10 +1,8 @@
+# Deploy configuration files
 - name: Remove input files example config files from rsyslog.d
+  vars:
+    __rsyslog_rules: "{{ rsyslog_files_input_rules }}"
```
@nhosoi You don't need __rsyslog_packages here?

nhosoi replied:
Well, we don't have a package cleanup code in cleanup.yaml. [1]  So, nothing happens even if we pass __rsyslog_packages...  Do you think we should uninstall packages?  If we uninstall rsyslog, it cleans up everything including the existing config files.  Is it preferable?  I checked linux-system-roles/{network,timesync,selinux}.  Network does not call "package"; the other 2 install packages with present, but I don't see absent...

[1] - https://github.com/linux-system-roles/logging/blob/master/roles/rsyslog/tasks/cleanup.yaml

sradco replied:
@nhosoi No. You are right. I missed that this is the cleanup.yaml

pcahyna replied:
FYI I don't think that we need an option to unistall the packages that we have installed. @nhosoi  As you observed, other roles don't do it either.

nhosoi replied:
Thanks for your input, @pcahyna.

sradco commented on roles/rsyslog/roles/input_roles/ovirt/tasks/main.yaml:
```diff
@@ -1,17 +1,16 @@
+# Deploy configuration files
 - name: Install/Update role packages and generate role configuration files to role subdir
+  vars:
+    __rsyslog_packages: "{{ __rsyslog_ovirt_prereq_packages | d([]) + __rsyslog_ovirt_packages | d([]) }}"
```
@nhosoi Why did you remove the "flatten"?

nhosoi replied:
Please see the comment https://github.com/linux-system-roles/logging/pull/87#discussion_r384707842

sradco commented on roles/rsyslog/roles/input_roles/viaq/tasks/main.yaml:
```diff
@@ -1,10 +1,9 @@
+# Deploy configuration files
 - name: Install/Update role packages and generate role configuration files to role subdir
+  vars:
+    __rsyslog_packages: "{{ __rsyslog_viaq_prereq_packages + __rsyslog_viaq_packages }}"
```
@nhosoi does this work like this?

nhosoi replied:
Yes.  I'm adding a ci test using viaq next...

richm replied:
so we can use this instead of `flatten` (re: above comments regarding `flatten`)?

nhosoi replied:
Yes, I double-checked by running my test with viaq, which requires 2 package lists and they are combined at the line 5.
```
vars/main.yml:__rsyslog_viaq_prereq_packages: ['nmap-ncat', 'systemd-python', 'policycoreutils', 'checkpolicy', 'policycoreutils-python']
vars/main.yml:__rsyslog_viaq_packages: ['rsyslog-mmnormalize']
```
I experimentally removed rpm packages 'nmap-ncat' and 'rsyslog-mmnormalize' on the target host and ran ansible-playbook to see these 2 packages are installed as expected.

As long as `__rsyslog_.*_prereq_packages` and `__rsylog_.*_packages` are the clean 1 level list, there's no need to have `flatten`.  And lists starting with __ are internal and we manage them, so it's safe to assume they are clean. (if not, it is a bug.)  First I thought having `flatten` is a reasonable safeguard, but I was told to remove them unless they are really necessary, most likely not to hide a hidden bugs.

richm replied:
I think even with user supplied values for packages that are supposed to be in `list` format, it is ok to use them with `__default_packages + user_supplied_packages`.  If the user does not provide `user_supplied_packages` in `list` format, they will get an error, which IMO is ok for now. We can go back later and add some sort of input validation if we feel that the default error is not sufficient for some reason.  Also, any such public values should be listed in `defaults/main.yml` and have a default value of `[]`.

nhosoi replied:
We already have a user supplied package list `rsyslog_extra_packages`.  It's initialized with [] in roles/rsyslog/defaults/main.yaml.
 roles/rsyslog/defaults/main.yaml
https://github.com/linux-system-roles/logging/pull/87/files#diff-def9bf9161e11520d46eea7aa85093e1R67-R70

And it's flattened before passing to the `package` module.  Hope it's ok...
https://github.com/linux-system-roles/logging/pull/87/files#diff-ed5a08a2e61430dc18e2d3820cbd6f32R3

sradco commented on roles/rsyslog/roles/output_roles/elasticsearch/tasks/main.yaml:
```diff
@@ -29,34 +29,35 @@
       when:
         - item.ca_cert_src is defined
-        - item.ca_cert_src != ""
+        - item.ca_cert_src
```
@nhosoi do they both mean the same
 item.ca_cert_src vs. item.ca_cert_src != ""   ?

nhosoi replied:
Yes.  This is part of adapting the ansible system roles best practices.

pcahyna replied:
@nhosoi are you sure? Also which part of the best practices are you referring to?

pcahyna replied:
see https://docs.ansible.com/ansible/latest/porting_guides/porting_guide_2.8.html#bare-variables-in-conditionals

nhosoi replied:
@sradco, @pcahyna, sorry, I obviously made a mistake.  Reverting this line.

pcahyna replied:
@sradco I am afraid they do not mean the same - this is a complicated subject, see the discussion in https://github.com/ansible/ansible-lint/issues/457 and related issues. Bare variables used to be treated differently than expressions in conditionals. I.e. `something` is not the same as `not (not something)`, even if evaluated in a boolean context.

pcahyna replied:
@nhosoi I think we should disable the problematic lint rule (E602), but there should be no problem with using `| length > 0` as well. Actually, I think now that `item.ca_cert_src` was actually correct, so @sradco was right - because it is not a bare variable - the dot in the middle makes it an expression. This is all so confusing :-(

pcahyna replied:
@nhosoi by the way, I just disabled the lint rule 602, which disallows comparisons to empty string, globally in the template. https://github.com/linux-system-roles/template/pull/19 . You can safely use comparison to empty string when you believe it is the best option. (After this change is merged.)

nhosoi replied:
Thank you for the update, @pcahyna !

pcahyna commented on .travis.yml:
```diff
@@ -10,7 +10,8 @@ env:
     - MOLECULE_DISTRO="fedora:31"
 install:
-  - pip install molecule docker
+  - pip uninstall molecule
```
@i386x @richm @tyll is the pip uninstall needed? Isn't `pip install 'molecule<3'` enough?

tyll replied:
the install command should be enough

nhosoi replied:
Thanks!  Removed uninstall.

pcahyna replied:
@nhosoi please rebase to split the travis change into a separate commit and squash it with the fixup here: https://github.com/linux-system-roles/logging/pull/87/commits/07e3fa0d7317a29c62eb5d237c41623c0cb97766#diff-354f30a63fb0907d4ad57269548329e3

nhosoi replied:
> @nhosoi please rebase to split the travis change into a separate commit and squash it with the fixup here: [07e3fa0#diff-354f30a63fb0907d4ad57269548329e3](https://github.com/linux-system-roles/logging/commit/07e3fa0d7317a29c62eb5d237c41623c0cb97766#diff-354f30a63fb0907d4ad57269548329e3)

@pcahyna, let me clarify your comment.  I was planning to squash 3 commits in this pr/#87 into one and merge.  Instead, are you suggesting to make 2 commits - one for travis changes and another for the rest?  Thanks.

pcahyna replied:
@nhosoi sorry, I did not realize that you want to squash all the commits. But yes, I think that the Travis changes should be in a separate commits from the rest. By the way, I think other changes may also stay in separate commits, like renaming the variables and then adjusting the filters - @richm what do you think?

nhosoi replied:
@pcahyna, it might be a bit difficult to make that many separate commits...  Travis is isolated, so it's easy.  For others, I'm afraid I may introduce some bugs (again...):  I don't trust myself. :p

pcahyna replied:
@nhosoi what I meant, after splitting off the Travis changes, you could leave the three commits as they are otherwise, making 4 commits in total. @sradco @richm would you prefer the existing commits, or all squashed in one except the Travis changes? By the way, you should still address https://github.com/linux-system-roles/logging/pull/87/commits/69140401652cca448b476c923fe7535f5989ea57#r387825151

nhosoi replied:
Ah, I see...  Thanks for the clarification, @pcahyna.
+1 to your suggestion.

richm replied:
> @sradco @richm would you prefer the existing commits, or all squashed in one except the Travis changes?

It depends. If @sradco needs to cherry pick these changes to the stable branch, then fewer commits is better.  Otherwise, I think 4 commits is fine.

nhosoi replied:
@richm, I don't think this pr is a candidate to backport to the stable branch except the travis/molecule change.  So, as long as the travis/molecule is isolated, we are good.

@sradco, do you want me to backport the travis fix to stable-0.1?

pcahyna commented on roles/rsyslog/defaults/main.yaml:
```diff
@@ -7,6 +7,11 @@
+#
+# true if the target host is a container and use rsyslog in the image.
+rsyslog_in_image: false
+
```
Is it a user-settable parameter? I don't see it in docs.

richm replied:
I think it is user-settable.  If `false` this means rsyslog is running on the host as a host package via systemd - the role needs to ensure that the rsyslog rpm packages are installed.  If `true`, this means rsyslog is running in a container - the role needs to ensure the right container is running - no rpm packages required.

nhosoi replied:
@richm, do you think this parameter should be in the doc to advertise?  Please note that the efforts we made in the openshift logging for rsyslog at the last moment are not reflected to this logging role.  Do you think they should be?

richm replied:
>  Do you think they should be?

Not sure, but let's worry about that in the future.  For now, for this PR, if `rsyslog_in_image` is settable by a user, and actual does something, let's doc it.

nhosoi replied:
Will do.  Thanks.

pcahyna commented on roles/rsyslog/tasks/deploy.yaml:
```diff
@@ -1,20 +1,23 @@
     state: latest
   when:
-    - not use_rsyslog_image | default(false) |bool
+    - not rsyslog_in_image | default(false) | bool
```
why `default(false)` if you have already set the default in defaults?

nhosoi replied:
Fixed.

pcahyna replied:
@nhosoi there are some remaining occurrences.

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -208,13 +215,13 @@
+        - not rsyslog_in_image | default(false) | bool
 
-  when: rsyslog_failed_validation | d(false) == false
+  when: not __rsyslog_failed_validation | d(false) | bool
```
do you need the bool filter here? This filter converts strings to booleans. Is the value supposed to be a string?

nhosoi replied:
Fixed.

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -208,13 +215,13 @@
+    msg: "Rsyslog version must be >= 8.37.0-7.2"
   when:
-    - rsyslog_failed_validation | d(false) == true
+    - __rsyslog_failed_validation | d(false) | bool
```
same here.

pcahyna commented on roles/rsyslog/tasks/main.yaml:
```diff
@@ -1,43 +1,34 @@
+    __rsyslog_version: >-
+      {{ ansible_facts.packages['rsyslog'][0].version
+      if ansible_facts.packages | selectattr('rsyslog', 'defined') | list | length > 0
+      else '0.0.0' }}
```
do you need set_fact or register? Can't you use vars: only on the block or tasks where the variables are actually used?

pcahyna replied:
I would say it is using `ansible_facts.packages`. But my comment was not about the way to obtain the version, but about what is being done with the value. I don't think a fact or registered variable should be used.

nhosoi replied:
@pcahyna, currently, __rsyslog_version is referred just in roles/rsyslog/tasks/main.yaml.  But I'm anticipating the value might be used in some other task files to absorb the version differences in the future.  What type of variable would be most appropriate for such a purpose?

richm replied:
> What type of variable would be most appropriate for such a purpose?

I'm not sure, but let's do that in a separate commit in the future, when such functionality is required.

nhosoi replied:
Ok.  So, is it ok to keep this change in this PR or should I revert it?
https://github.com/linux-system-roles/logging/pull/87/files#diff-ed5a08a2e61430dc18e2d3820cbd6f32L14-R17

richm replied:
Hmm - for some reason that link isn't working for me - what is that change?

nhosoi replied:
Sorry, this is the raw diff.  (indeed, the link does not work for me, either ):
Actually, I realized there were 2 issues.  In the beginning, rsyslog version was retrieved using the yum module [1].  But we are to use `package` which is a higher level, platform dependent module.  Unfortunately, it does not return version.  So, I fixed it with the ansible_facts data [2].  That being said, we do not want to go back to `yum:`?

[1]
```
 - name: Get rsyslog version
-  yum:
+  package:
     list: rsyslog
   register: rsyslog_version
```
[2]
```
 - name: Get rsyslog version
-  package:
-    list: rsyslog
-  register: rsyslog_version
+  set_fact:
+    __rsyslog_version: >-
+      {{ ansible_facts.packages['rsyslog'][0].version
+      if ansible_facts.packages | selectattr('rsyslog', 'defined') | list | length > 0
+      else '0.0.0' }}
```

richm replied:
no, we do not want to go back to `yum` 
as pcahyna said, using `ansible_facts.packages['rsyslog'][0].version` is the canonical way to get the package version.

pcahyna replied:
By the way, do you really need `| length > 0` ?

nhosoi replied:
> By the way, do you really need `| length > 0` ?

@pcahyna, no, it's not needed.

nhosoi replied:
@pchayna
> By the way, do you really need `| length > 0` ?

I've opened a pr to remove `| length > 0`.  Could you please take a look?
https://github.com/linux-system-roles/logging/pull/91
Also, it turned out selectattr did not work for this case [1], thus I replaced it with json_query.  But if you think I should use the timesync method https://github.com/linux-system-roles/timesync/pull/60, please add your comment to https://github.com/linux-system-roles/logging/pull/91.  I'm going to update the pr.  Thanks.

[1] - Debug outputs from `ansible_facts.packages | json_query('rsyslog')` and `ansible_facts.packages | selectattr('rsyslog', 'defined') | list`.
```
TASK [linux-system-roles/logging/roles/rsyslog : debug] *******************************************
task path: linux-system-roles/logging/roles/rsyslog/tasks/main.yaml:12
ok: [momo7.localdomain] => {
    "msg": "json_query - [{'name': 'rsyslog', 'source': 'rpm', 'epoch': None, 'version': '8.24.0', 'release': '41.el7_7.2', 'arch': 'x86_64'}]"
}

TASK [linux-system-roles/logging/roles/rsyslog : debug] *******************************************
task path: linux-system-roles/logging/roles/rsyslog/tasks/main.yaml:15
ok: [momo7.localdomain] => {
    "msg": "selectattr - []"
}
```

pcahyna replied:
maybe I already said this but timesync now uses an even simpler approach: https://github.com/linux-system-roles/timesync/pull/67/files#diff-6d62b73b3f46cfa11f6466d08cf45303

richm commented on roles/rsyslog/README.md:
````diff
@@ -338,11 +338,11 @@ WARNING: Pre-existing rsyslog.conf and configuration files in /etc/rsyslog.d are
+To make a new configuration file installed in addition to the current {{ __rsyslog_viaq_rules }}, create an rsyslog config item based on the following skelton and add the title {{ __rsyslog_conf_yourname }} to {{ __rsyslog_viaq_rules }}.
 ```
-rsyslog_conf_yourname:
+__rsyslog_conf_yourname:
````
Are these instructions intended for a role developer e.g. to add a new default configuration file to the role repo?  If so, then ok.

nhosoi replied:
Yes, I believe so.  I'm adding "to add a new default configuration file to the role repo" to the doc.  Thanks.

richm commented on roles/rsyslog/roles/input_roles/files/defaults/main.yaml:
```diff
@@ -68,7 +68,7 @@ rsyslog_conf_imfile_inputs:
         options: |-
           # FIXME: Need to normalize a log before passing it to elasticsearch output
-          {% if rsyslog_elasticsearch is defined and rsyslog_elasticsearch|length > 0 %}
+          {% if rsyslog_elasticsearch is defined and rsyslog_elasticsearch | length > 0 %}
```
if `rsyslog_elasticsearch` is a `list`, `dict`, or `string`, you do not need to test the length.  https://docs.ansible.com/ansible/latest/user_guide/playbooks_tests.html and in particular https://jinja.palletsprojects.com/en/2.11.x/templates/#if - you can just use `if rsyslog_elasticsearch is defined and rsyslog_elasticsearch` - this means `if rsyslog_elasticsearch is defined and rsyslog_elasticsearch is not empty`

richm commented:
lgtm - pcahyna?

nhosoi commented:
Thank you all of you who reviewed this pr and answered to the questions.

Merging the pr.
Note: I created a separate commit from the travis/molecule change [3] and another commit addressing the reviews [2] (it used to be 2 commits which are merged into one [2]).  As shown in the commits page, 3 commits are involved.
https://github.com/linux-system-roles/logging/pull/87/commits
[1] Fixing package module related issues.
[2] Fixes to address the issues found in the review process.
[3] Workaround for the molecule3 upgrade - forcing to install molecule < 3.
